bark_ml/observers/graph_observer.py

```diff
@@ -0,0 +1,554 @@
+    self._normalize_observations = normalize_observations
+    self._logger = logging.getLogger()
+
+    self._agent_limit =\
+      params["ML"]["GraphObserver"]["AgentLimit", 
+      "The maximum number of agents that can be observed.",
+      4]
+
+    self._visibility_radius =\
+      params["ML"]["GraphObserver"]["VisibilityRadius",
+      "The radius in which agent can 'see', i.e. detect other agents.", 
+      50]
+
+    self._add_self_loops =\
+      params["ML"]["GraphObserver"]["SelfLoops", 
+      "Whether each node has an edge pointing to itself.", 
+      False]
+
+    requested_node_attribute_keys =\
+      params["ML"]["GraphObserver"]["EnabledNodeFeatures",
+      "The list of available node features, given by their string key that \
+       the observer should extract from the world and insert into the \
+       observation. For a list of available features, refer to the list \
+       returned by `GraphObserver.available_node_attributes`.",
+      self.available_node_attributes()
+    ]
+
+    self.enabled_node_attribute_keys = self._filter_requested_attributes(
+      requested_keys=requested_node_attribute_keys,
+      available_keys=self.available_node_attributes(),
+      context="node")
+    self._logger.info(
+      "GraphObserver configured with node attributes: " +
+      f"{self.enabled_node_attribute_keys}")
+
+    requested_edge_attribute_keys =\
+      params["ML"]["GraphObserver"]["EnabledEdgeFeatures",
+      "The list of available edge features, given by their string key that \
+       the observer should extract from the world and insert into the \
+       observation. For a list of available features, refer to the list \
+       returned by `GraphObserver.available_edge_attributes`.",
+      self.available_edge_attributes()
+    ]
+
+    self.enabled_edge_attribute_keys = self._filter_requested_attributes(
+      requested_keys=requested_edge_attribute_keys,
+      available_keys=self.available_edge_attributes(),
+      context="edge")
+    self._logger.info(
+      "GraphObserver configured with edge attributes: " +
+      f"{self.enabled_edge_attribute_keys}")
+
+    # the number of features of a node in the graph
+    self.feature_len = len(self.enabled_node_attribute_keys)
+
+    # the number of features of an edge between two nodes
+    self.edge_feature_len = len(self.enabled_edge_attribute_keys)
+
+  def Observe(self, world):
+    """see base class"""
+    agents = self._preprocess_agents(world)
+
+    # placeholder for the output observation
+    obs = np.zeros(self._len_state)
+
+    # insert node features for each agent
+    for i, agent in agents:
+      start_index = i * self.feature_len
+      end_index = start_index + self.feature_len
+      obs[start_index:end_index] = self._extract_node_features(agent)
+
+    edge_features = np.zeros((self._agent_limit, self._agent_limit, self.edge_feature_len))
+    adjacency_matrix = np.zeros((self._agent_limit, self._agent_limit))
```

Comment:
does this imply the adjacency_matrix only is e.g. 4x4 (as this is set as the agent limit)?
If there are 4 agents, each agent can be connected with 3 other agents assuming directed edges + self connections.
So should the adjacency_matrix not be of larger size (and of course the edge features)?

Reply:
> does this imply the adjacency_matrix only is e.g. 4x4 (as this is set as the agent limit)?
If there are 4 agents, each agent can be connected with 3 other agents assuming directed edges + self connections.
Yes, that's correct.

> So should the adjacency_matrix not be of larger size (and of course the edge features)?
What makes you think that? 
Assuming 4 agents, the matrix contains a row for each agent (-> 4 rows) and each row contains a binary column for each agent (-> 4 columns) indicating whether an edge to this agent exists. Hence, a 4x4 matrix.

Reply:
ah okay, well was quite late yesterday!